src/trap.S

```diff
@@ -94,3 +110,68 @@ _trap_entry:
     ld   sp, ({tf_sp})(sp)
 
     sret
+
+_stack_overflow:
+    /* Acquire the emergency stack on a spinlock. See "A spinlock with fences" in [1].
+     *
+     * [1]: RISC-V Memory Consistency Model Specification (DRAFT), Dec 1st 2017, 2.3.5.
+     */
+    la t1, _emergency_stack_lock
+    li t0, 1
+1:
+    amoswap.w t0, t0, (t1)
+    fence r, rw
+    bnez t0, 1b
+
+    csrr t0, sscratch
+    /* NOTE: Register T1 will be clobbered in this stack frame.*/
+    csrw sscratch, sp
+    la sp, _emergency_stack_top
+    sd   ra, ({tf_ra})(sp)
+    sd   gp, ({tf_gp})(sp)
+    sd   tp, ({tf_tp})(sp)
+    sd   s0, ({tf_s0})(sp)
+    sd   s1, ({tf_s1})(sp)
+    sd   a0, ({tf_a0})(sp)
+    sd   a1, ({tf_a1})(sp)
+    sd   a2, ({tf_a2})(sp)
+    sd   a3, ({tf_a3})(sp)
+    sd   a4, ({tf_a4})(sp)
+    sd   a5, ({tf_a5})(sp)
+    sd   a6, ({tf_a6})(sp)
+    sd   a7, ({tf_a7})(sp)
+    sd   s2, ({tf_s2})(sp)
+    sd   s3, ({tf_s3})(sp)
+    sd   s4, ({tf_s4})(sp)
+    sd   s5, ({tf_s5})(sp)
+    sd   s6, ({tf_s6})(sp)
+    sd   s7, ({tf_s7})(sp)
+    sd   s8, ({tf_s8})(sp)
+    sd   s9, ({tf_s9})(sp)
+    sd   s10, ({tf_s10})(sp)
+    sd   s11, ({tf_s11})(sp)
+    sd   t0, ({tf_t0})(sp)
+    sd   t1, ({tf_t1})(sp)
+    sd   t2, ({tf_t2})(sp)
+    sd   t3, ({tf_t3})(sp)
+    sd   t4, ({tf_t4})(sp)
+    sd   t5, ({tf_t5})(sp)
+    sd   t6, ({tf_t6})(sp)
+    csrr t0, sscratch
+    sd   t0, ({tf_sp})(sp)
+    csrw sscratch, zero
+    csrr t1, sstatus
+    sd   t1, ({tf_sstatus})(sp)
+    csrr t2, sepc
+    sd   t2, ({tf_sepc})(sp)
+
+    /* Now enter the rust trap handler. */
+    move a0, sp
+    call handle_stack_overflow
+
+.data
+_emergency_stack_lock:
+.long 0
+_emergency_stack_bottom:
+ .space 64*1024
+_emergency_stack_top:
```

Comment:
can we declare these in rust? and do we actually need 16k here?

Reply:
Yes, will do. No, I am just being paranoid in stack usage, will reduce.

Reply:
I could declare stack_lock in rust. Regarding the overflow stack itself, it was a bit hairy to get the pointer to the top of the stack. I went for definition of the overflow stack in linker script.